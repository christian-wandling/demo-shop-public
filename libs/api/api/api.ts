export * from './cart-items-api';
import { CartItemsApi } from './cart-items-api';
export * from './orders-api';
import { OrdersApi } from './orders-api';
export * from './products-api';
import { ProductsApi } from './products-api';
export * from './shopping-sessions-api';
import { ShoppingSessionsApi } from './shopping-sessions-api';
export * from './users-api';
import { UsersApi } from './users-api';
export const APIS = [CartItemsApi, OrdersApi, ProductsApi, ShoppingSessionsApi, UsersApi];
