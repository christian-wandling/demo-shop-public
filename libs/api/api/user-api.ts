/**
 * Demo Shop API
 *
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
/* tslint:disable:no-unused-variable member-ordering */

import { Inject, Injectable, Optional } from '@angular/core';
import {
  HttpClient,
  HttpHeaders,
  HttpParams,
  HttpResponse,
  HttpEvent,
  HttpParameterCodec,
  HttpContext,
} from '@angular/common/http';
import { CustomHttpParameterCodec } from '../encoder';
import { Observable } from 'rxjs';

// @ts-ignore
import { AddressResponse } from '../model/address-response';
// @ts-ignore
import { UpdateUserAddressRequest } from '../model/update-user-address-request';
// @ts-ignore
import { UpdateUserPhoneRequest } from '../model/update-user-phone-request';
// @ts-ignore
import { UserResponse } from '../model/user-response';

// @ts-ignore
import { BASE_PATH, COLLECTION_FORMATS } from '../variables';
import { Configuration } from '../configuration';

@Injectable({
  providedIn: 'root',
})
export class UserApi {
  protected basePath = 'http://localhost';
  public defaultHeaders = new HttpHeaders();
  public configuration = new Configuration();
  public encoder: HttpParameterCodec;

  constructor(
    protected httpClient: HttpClient,
    @Optional() @Inject(BASE_PATH) basePath: string | string[],
    @Optional() configuration: Configuration
  ) {
    if (configuration) {
      this.configuration = configuration;
    }
    if (typeof this.configuration.basePath !== 'string') {
      const firstBasePath = Array.isArray(basePath) ? basePath[0] : undefined;
      if (firstBasePath != undefined) {
        basePath = firstBasePath;
      }

      if (typeof basePath !== 'string') {
        basePath = this.basePath;
      }
      this.configuration.basePath = basePath;
    }
    this.encoder = this.configuration.encoder || new CustomHttpParameterCodec();
  }

  // @ts-ignore
  private addToHttpParams(httpParams: HttpParams, value: any, key?: string): HttpParams {
    if (typeof value === 'object' && value instanceof Date === false) {
      httpParams = this.addToHttpParamsRecursive(httpParams, value);
    } else {
      httpParams = this.addToHttpParamsRecursive(httpParams, value, key);
    }
    return httpParams;
  }

  private addToHttpParamsRecursive(httpParams: HttpParams, value?: any, key?: string): HttpParams {
    if (value == null) {
      return httpParams;
    }

    if (typeof value === 'object') {
      if (Array.isArray(value)) {
        (value as any[]).forEach(elem => (httpParams = this.addToHttpParamsRecursive(httpParams, elem, key)));
      } else if (value instanceof Date) {
        if (key != null) {
          httpParams = httpParams.append(key, (value as Date).toISOString().substring(0, 10));
        } else {
          throw Error('key may not be null if value is Date');
        }
      } else {
        Object.keys(value).forEach(
          k => (httpParams = this.addToHttpParamsRecursive(httpParams, value[k], key != null ? `${key}.${k}` : k))
        );
      }
    } else if (key != null) {
      httpParams = httpParams.append(key, value);
    } else {
      throw Error('key may not be null if value is not object or array');
    }
    return httpParams;
  }

  /**
   * Resolve current user
   * Resolve current user based on identity extracted from bearer token
   * @param observe set whether or not to return the data Observable as the body, response or events. defaults to returning the body.
   * @param reportProgress flag to report request and response progress.
   */
  public resolveCurrentUser(
    observe?: 'body',
    reportProgress?: boolean,
    options?: { httpHeaderAccept?: 'application/json'; context?: HttpContext; transferCache?: boolean }
  ): Observable<UserResponse>;
  public resolveCurrentUser(
    observe?: 'response',
    reportProgress?: boolean,
    options?: { httpHeaderAccept?: 'application/json'; context?: HttpContext; transferCache?: boolean }
  ): Observable<HttpResponse<UserResponse>>;
  public resolveCurrentUser(
    observe?: 'events',
    reportProgress?: boolean,
    options?: { httpHeaderAccept?: 'application/json'; context?: HttpContext; transferCache?: boolean }
  ): Observable<HttpEvent<UserResponse>>;
  public resolveCurrentUser(
    observe: any = 'body',
    reportProgress: boolean = false,
    options?: { httpHeaderAccept?: 'application/json'; context?: HttpContext; transferCache?: boolean }
  ): Observable<any> {
    let localVarHeaders = this.defaultHeaders;

    let localVarCredential: string | undefined;
    // authentication (bearer) required
    localVarCredential = this.configuration.lookupCredential('bearer');
    if (localVarCredential) {
      localVarHeaders = localVarHeaders.set('Authorization', 'Bearer ' + localVarCredential);
    }

    let localVarHttpHeaderAcceptSelected: string | undefined = options && options.httpHeaderAccept;
    if (localVarHttpHeaderAcceptSelected === undefined) {
      // to determine the Accept header
      const httpHeaderAccepts: string[] = ['application/json'];
      localVarHttpHeaderAcceptSelected = this.configuration.selectHeaderAccept(httpHeaderAccepts);
    }
    if (localVarHttpHeaderAcceptSelected !== undefined) {
      localVarHeaders = localVarHeaders.set('Accept', localVarHttpHeaderAcceptSelected);
    }

    let localVarHttpContext: HttpContext | undefined = options && options.context;
    if (localVarHttpContext === undefined) {
      localVarHttpContext = new HttpContext();
    }

    let localVarTransferCache: boolean | undefined = options && options.transferCache;
    if (localVarTransferCache === undefined) {
      localVarTransferCache = true;
    }

    let responseType_: 'text' | 'json' | 'blob' = 'json';
    if (localVarHttpHeaderAcceptSelected) {
      if (localVarHttpHeaderAcceptSelected.startsWith('text')) {
        responseType_ = 'text';
      } else if (this.configuration.isJsonMime(localVarHttpHeaderAcceptSelected)) {
        responseType_ = 'json';
      } else {
        responseType_ = 'blob';
      }
    }

    let localVarPath = `/api/v1/users/me`;
    return this.httpClient.request<UserResponse>('post', `${this.configuration.basePath}${localVarPath}`, {
      context: localVarHttpContext,
      responseType: <any>responseType_,
      withCredentials: this.configuration.withCredentials,
      headers: localVarHeaders,
      observe: observe,
      transferCache: localVarTransferCache,
      reportProgress: reportProgress,
    });
  }

  /**
   * Update address of current user
   * Update the address of the user based on identity extracted from bearer token
   * @param updateUserAddressRequest
   * @param observe set whether or not to return the data Observable as the body, response or events. defaults to returning the body.
   * @param reportProgress flag to report request and response progress.
   */
  public updateCurrentUserAddress(
    updateUserAddressRequest: UpdateUserAddressRequest,
    observe?: 'body',
    reportProgress?: boolean,
    options?: { httpHeaderAccept?: 'application/json'; context?: HttpContext; transferCache?: boolean }
  ): Observable<AddressResponse>;
  public updateCurrentUserAddress(
    updateUserAddressRequest: UpdateUserAddressRequest,
    observe?: 'response',
    reportProgress?: boolean,
    options?: { httpHeaderAccept?: 'application/json'; context?: HttpContext; transferCache?: boolean }
  ): Observable<HttpResponse<AddressResponse>>;
  public updateCurrentUserAddress(
    updateUserAddressRequest: UpdateUserAddressRequest,
    observe?: 'events',
    reportProgress?: boolean,
    options?: { httpHeaderAccept?: 'application/json'; context?: HttpContext; transferCache?: boolean }
  ): Observable<HttpEvent<AddressResponse>>;
  public updateCurrentUserAddress(
    updateUserAddressRequest: UpdateUserAddressRequest,
    observe: any = 'body',
    reportProgress: boolean = false,
    options?: { httpHeaderAccept?: 'application/json'; context?: HttpContext; transferCache?: boolean }
  ): Observable<any> {
    if (updateUserAddressRequest === null || updateUserAddressRequest === undefined) {
      throw new Error(
        'Required parameter updateUserAddressRequest was null or undefined when calling updateCurrentUserAddress.'
      );
    }

    let localVarHeaders = this.defaultHeaders;

    let localVarCredential: string | undefined;
    // authentication (bearer) required
    localVarCredential = this.configuration.lookupCredential('bearer');
    if (localVarCredential) {
      localVarHeaders = localVarHeaders.set('Authorization', 'Bearer ' + localVarCredential);
    }

    let localVarHttpHeaderAcceptSelected: string | undefined = options && options.httpHeaderAccept;
    if (localVarHttpHeaderAcceptSelected === undefined) {
      // to determine the Accept header
      const httpHeaderAccepts: string[] = ['application/json'];
      localVarHttpHeaderAcceptSelected = this.configuration.selectHeaderAccept(httpHeaderAccepts);
    }
    if (localVarHttpHeaderAcceptSelected !== undefined) {
      localVarHeaders = localVarHeaders.set('Accept', localVarHttpHeaderAcceptSelected);
    }

    let localVarHttpContext: HttpContext | undefined = options && options.context;
    if (localVarHttpContext === undefined) {
      localVarHttpContext = new HttpContext();
    }

    let localVarTransferCache: boolean | undefined = options && options.transferCache;
    if (localVarTransferCache === undefined) {
      localVarTransferCache = true;
    }

    // to determine the Content-Type header
    const consumes: string[] = ['application/json'];
    const httpContentTypeSelected: string | undefined = this.configuration.selectHeaderContentType(consumes);
    if (httpContentTypeSelected !== undefined) {
      localVarHeaders = localVarHeaders.set('Content-Type', httpContentTypeSelected);
    }

    let responseType_: 'text' | 'json' | 'blob' = 'json';
    if (localVarHttpHeaderAcceptSelected) {
      if (localVarHttpHeaderAcceptSelected.startsWith('text')) {
        responseType_ = 'text';
      } else if (this.configuration.isJsonMime(localVarHttpHeaderAcceptSelected)) {
        responseType_ = 'json';
      } else {
        responseType_ = 'blob';
      }
    }

    let localVarPath = `/api/v1/users/me/address`;
    return this.httpClient.request<AddressResponse>('put', `${this.configuration.basePath}${localVarPath}`, {
      context: localVarHttpContext,
      body: updateUserAddressRequest,
      responseType: <any>responseType_,
      withCredentials: this.configuration.withCredentials,
      headers: localVarHeaders,
      observe: observe,
      transferCache: localVarTransferCache,
      reportProgress: reportProgress,
    });
  }

  /**
   * Update phone of current user
   * Update the phone of the user based on identity extracted from bearer token
   * @param updateUserPhoneRequest
   * @param observe set whether or not to return the data Observable as the body, response or events. defaults to returning the body.
   * @param reportProgress flag to report request and response progress.
   */
  public updateCurrentUserPhone(
    updateUserPhoneRequest: UpdateUserPhoneRequest,
    observe?: 'body',
    reportProgress?: boolean,
    options?: { httpHeaderAccept?: 'application/json'; context?: HttpContext; transferCache?: boolean }
  ): Observable<UserResponse>;
  public updateCurrentUserPhone(
    updateUserPhoneRequest: UpdateUserPhoneRequest,
    observe?: 'response',
    reportProgress?: boolean,
    options?: { httpHeaderAccept?: 'application/json'; context?: HttpContext; transferCache?: boolean }
  ): Observable<HttpResponse<UserResponse>>;
  public updateCurrentUserPhone(
    updateUserPhoneRequest: UpdateUserPhoneRequest,
    observe?: 'events',
    reportProgress?: boolean,
    options?: { httpHeaderAccept?: 'application/json'; context?: HttpContext; transferCache?: boolean }
  ): Observable<HttpEvent<UserResponse>>;
  public updateCurrentUserPhone(
    updateUserPhoneRequest: UpdateUserPhoneRequest,
    observe: any = 'body',
    reportProgress: boolean = false,
    options?: { httpHeaderAccept?: 'application/json'; context?: HttpContext; transferCache?: boolean }
  ): Observable<any> {
    if (updateUserPhoneRequest === null || updateUserPhoneRequest === undefined) {
      throw new Error(
        'Required parameter updateUserPhoneRequest was null or undefined when calling updateCurrentUserPhone.'
      );
    }

    let localVarHeaders = this.defaultHeaders;

    let localVarCredential: string | undefined;
    // authentication (bearer) required
    localVarCredential = this.configuration.lookupCredential('bearer');
    if (localVarCredential) {
      localVarHeaders = localVarHeaders.set('Authorization', 'Bearer ' + localVarCredential);
    }

    let localVarHttpHeaderAcceptSelected: string | undefined = options && options.httpHeaderAccept;
    if (localVarHttpHeaderAcceptSelected === undefined) {
      // to determine the Accept header
      const httpHeaderAccepts: string[] = ['application/json'];
      localVarHttpHeaderAcceptSelected = this.configuration.selectHeaderAccept(httpHeaderAccepts);
    }
    if (localVarHttpHeaderAcceptSelected !== undefined) {
      localVarHeaders = localVarHeaders.set('Accept', localVarHttpHeaderAcceptSelected);
    }

    let localVarHttpContext: HttpContext | undefined = options && options.context;
    if (localVarHttpContext === undefined) {
      localVarHttpContext = new HttpContext();
    }

    let localVarTransferCache: boolean | undefined = options && options.transferCache;
    if (localVarTransferCache === undefined) {
      localVarTransferCache = true;
    }

    // to determine the Content-Type header
    const consumes: string[] = ['application/json'];
    const httpContentTypeSelected: string | undefined = this.configuration.selectHeaderContentType(consumes);
    if (httpContentTypeSelected !== undefined) {
      localVarHeaders = localVarHeaders.set('Content-Type', httpContentTypeSelected);
    }

    let responseType_: 'text' | 'json' | 'blob' = 'json';
    if (localVarHttpHeaderAcceptSelected) {
      if (localVarHttpHeaderAcceptSelected.startsWith('text')) {
        responseType_ = 'text';
      } else if (this.configuration.isJsonMime(localVarHttpHeaderAcceptSelected)) {
        responseType_ = 'json';
      } else {
        responseType_ = 'blob';
      }
    }

    let localVarPath = `/api/v1/users/me/phone`;
    return this.httpClient.request<UserResponse>('patch', `${this.configuration.basePath}${localVarPath}`, {
      context: localVarHttpContext,
      body: updateUserPhoneRequest,
      responseType: <any>responseType_,
      withCredentials: this.configuration.withCredentials,
      headers: localVarHeaders,
      observe: observe,
      transferCache: localVarTransferCache,
      reportProgress: reportProgress,
    });
  }
}
