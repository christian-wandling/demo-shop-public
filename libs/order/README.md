# `@demo-shop/order`

This library handles aspects of the order feature including order selection and invoice pdf generation.

## Features

- **Order Management**

  - List orders history of current user

- **Order Details**

  - Comprehensive order information display
  - Product details with images and pricing

- **Invoice Generation**

  - In-browser PDF generation and download

- **State Management**
  - Signal-based reactive state
  - Synchronization with backend

## API Reference

For a complete API reference, see the [Compodoc documentation](/apps/frontend/docs/compodoc).

## Running unit tests

Run `nx test order` to execute the unit tests.
