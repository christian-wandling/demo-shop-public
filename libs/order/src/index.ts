export * from './lib/order.routes';
export * from './lib/order.facade';
