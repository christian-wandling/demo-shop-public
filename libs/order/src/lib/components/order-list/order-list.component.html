<div class="mx-auto px-8 max-w-7xl">
  <section class="py-8 top-16 sticky z-10 bg-white border-b-2 border-gray-100">
    @if (user()) {
    <h1 class="font-bold text-gray-700" data-testid="order-list-username">
      {{ user()?.firstname }} {{ user()?.lastname }}
    </h1>
    <div class="text-xs text-gray-500">
      <p data-testid="order-list-address1">{{ user()?.address?.street }} {{ user()?.address?.apartment }}</p>
      <p data-testid="order-list-address2">
        {{ user()?.address?.zip }} {{ user()?.address?.city }} , {{ user()?.address?.region }}
      </p>
      <p data-testid="order-list-address3">{{ user()?.address?.country }}</p>
    </div>
    }
  </section>
  <section>
    <ul class="divide-y divide-gray-100">
      @for (order of orders(); track order.id) {
      <li
        class="relative flex flex-col md:flex-row justify-between gap-x-3 sm:gap-x-6 py-2 sm:py-5 py-2 sm:px-5 cursor-pointer hover:bg-slate-50"
        routerLink="{{ order.id }}"
        [attr.data-testid]="'order-list-item-link' + order.id">
        <div class="flex min-w-0 gap-x-3 mb-4">
          <div class="min-w-0 flex-auto">
            <p class="text-xl mb-4 font-semibold hover:underline" [attr.data-testid]="'order-list-item-id' + order.id">
              Order #{{ order.id }}
            </p>
            <div class="flex gap-1 sm:gap-4">
              @for (item of order.items; track item.productId; let index = $index; let count = $count) { @if (index <=
              MAX_THUMBNAILS - 1) { @if (index === MAX_THUMBNAILS - 1 && count > MAX_THUMBNAILS) {
              <div
                class="w-12 h-12 sm:w-24 sm:h-24 aspect-1 flex-shrink-0 bg-gray-800 text-xl sm:text-5xl text-white flex items-center justify-center rounded-lg"
                [attr.data-testid]="'order-list-item-thumbnails-more' + order.id">
                +{{ count - MAX_THUMBNAILS + 1 }}
              </div>
              } @else {
              <div
                class="relative w-12 h-12 sm:w-24 sm:h-24 aspect-1 flex-shrink-0 rounded-lg overflow-hidden group-hover:opacity-75"
                routerLink="/products/{{ item.productId }}"
                [attr.data-testid]="'order-list-item-product-link' + order.id + '-' + item.productId">
                <img
                  [ngSrc]="item.productThumbnail"
                  [alt]="item.productName"
                  [title]="item.productName"
                  [attr.data-testid]="'order-list-item-thumbnails' + order.id + '-' + item.productId"
                  placeholder
                  fill="" />
              </div>
              } } }
            </div>
          </div>
        </div>
        <lib-order-status
          [status]="order.status"
          class="sm:absolute sm:right-5 sm:top-5 mb-4"
          [attr.data-testid]="'order-list-item-status' + order.id" />
        <div class="flex flex-col md:items-end sm:justify-end gap-1">
          <p class="text-sm leading-none text-gray-900" [attr.data-testid]="'order-list-item-amount' + order.id">
            {{ order.amount | currency }}
          </p>
          <lib-date-time
            [dateTime]="order.created"
            class="text-xs leading-none text-gray-500"
            [attr.data-testid]="'order-list-item-date' + order.id" />
        </div>
      </li>
      }
    </ul>
  </section>
</div>
