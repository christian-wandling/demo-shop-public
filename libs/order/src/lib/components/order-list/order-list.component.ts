import { ChangeDetectionStrategy, Component, inject, OnInit } from '@angular/core';
import { CommonModule, NgOptimizedImage } from '@angular/common';
import { OrderFacade } from '../../order.facade';
import { UserFacade } from '@demo-shop/user';
import { RouterLink } from '@angular/router';
import { OrderStatusComponent } from '../shared/order-status/order-status.component';
import { DateTimeComponent } from '@demo-shop/shared';

/**
 * Component that displays a list of orders for the current user.
 * The orders are sorted by status and date.
 *
 * @example
 * <lib-order-list/>
 */
@Component({
  selector: 'lib-order-list',
  standalone: true,
  imports: [CommonModule, RouterLink, NgOptimizedImage, OrderStatusComponent, DateTimeComponent],
  templateUrl: './order-list.component.html',
  styleUrl: './order-list.component.scss',
  changeDetection: ChangeDetectionStrategy.OnPush,
  host: {
    'data-testid': 'order-list',
  },
})
export class OrderListComponent implements OnInit {
  /**
   * Maximum number of product thumbnails to display per order
   */
  readonly MAX_THUMBNAILS = 5;

  readonly #userFacade = inject(UserFacade);
  readonly user = this.#userFacade.getCurrentUser();
  readonly #orderFacade = inject(OrderFacade);
  readonly orders = this.#orderFacade.getSortedByStatusAndDate();

  /**
   * Fetches all orders for the current user
   */
  ngOnInit() {
    this.#orderFacade.fetchAll();
  }
}
