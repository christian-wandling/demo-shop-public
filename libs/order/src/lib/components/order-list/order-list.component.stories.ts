import { applicationConfig, Meta, StoryObj } from '@storybook/angular';
import { CommonModule, NgOptimizedImage } from '@angular/common';
import { provideRouter } from '@angular/router';
import { importProvidersFrom, signal } from '@angular/core';
import { AddressResponse, OrderItemResponse, OrderResponse, OrderStatus, UserResponse } from '@demo-shop/api';
import { within } from '@storybook/testing-library';
import { expect } from '@storybook/jest';
import { faker } from '@faker-js/faker';
import { provideImageLoader } from '@demo-shop/shared';
import { OrderListComponent } from './order-list.component';
import { OrderFacade } from '../../order.facade';
import { UserFacade } from '@demo-shop/user';

const orders: OrderResponse[] = Array.from({ length: 10 }).map((_, id) => {
  const items: OrderItemResponse[] = Array.from({
    length: faker.number.int({
      min: 1,
      max: 10,
    }),
  }).map((_, productId) => {
    const quantity = faker.number.int({ min: 1, max: 5 });
    const unitPrice = faker.number.int({ min: 1, max: 20 }) * 10 - 1;

    return {
      productId,
      productName: faker.commerce.productName(),
      productThumbnail: faker.image.urlPicsumPhotos(),
      quantity,
      unitPrice,
      totalPrice: quantity * unitPrice,
    };
  });

  return {
    id,
    amount: items.reduce((acc, curr) => acc + curr.totalPrice, 0),
    created: faker.date.past().toISOString(),
    items,
    status: Object.values(OrderStatus)[id % 2],
    userId: 0,
  };
});

const address: AddressResponse = {
  street: faker.location.streetAddress(),
  apartment: faker.number.int({ max: 100 }).toString(),
  city: faker.location.city(),
  zip: faker.location.zipCode('#####'),
  region: faker.location.state(),
  country: faker.location.country(),
};

const user: UserResponse = {
  id: 1,
  email: faker.internet.email(),
  firstname: faker.person.firstName(),
  lastname: faker.person.lastName(),
  phone: faker.helpers.fromRegExp('+[0-9]{9}'),
  address,
};

const mockOrderFacade = {
  getSortedByStatusAndDate: () => {
    orders.sort((a, b) => {
      if (a.status === b.status) {
        return a.created > b.created ? -1 : 1;
      }

      return a.status === OrderStatus.Created ? -1 : 1;
    });

    return signal(orders);
  },
  fetchAll: () => {
    return;
  },
};

const mockUserFacade = {
  getCurrentUser: () => signal(user),
};

const meta: Meta<OrderListComponent> = {
  component: OrderListComponent,
  title: 'Order/OrderListComponent',
  decorators: [
    applicationConfig({
      providers: [
        importProvidersFrom(CommonModule, NgOptimizedImage),
        provideImageLoader(),
        provideRouter([{ path: '**', redirectTo: '' }]),
        { provide: OrderFacade, useValue: mockOrderFacade },
        { provide: UserFacade, useValue: mockUserFacade },
      ],
    }),
  ],
  parameters: {
    controls: {
      exclude: ['#orderFacade', '#userFacade', 'orders', 'user', 'ngOnInit'],
    },
  },
};
export default meta;
type Story = StoryObj<OrderListComponent>;

export const Default: Story = {
  render: args => ({
    props: args,
    template: `
      <div>
        <div style="height: 4rem; width: 100%; background: #fff; z-index: 10; top: 0; position: sticky"></div>
        <lib-order-list></lib-order-list>
      </div>
    `,
  }),
  play: async ({ canvasElement, args }) => {
    const canvas = within(canvasElement);
    const name = canvas.getByText(`${user.firstname} ${user.lastname}`);
    const addressLine1 = canvas.getByText(`${address.street} ${address.apartment}`);
    const addressLine2 = canvas.getByText(`${address.zip} ${address.city}, ${address.region}`);
    const addressLine3 = canvas.getByText(`${address.country}`);
    const orders = canvas.getAllByText('Order #');
    expect(name).toBeTruthy();
    expect(addressLine1).toBeTruthy();
    expect(addressLine2).toBeTruthy();
    expect(addressLine3).toBeTruthy();
    expect(orders).toHaveLength(orders.length);
  },
};
