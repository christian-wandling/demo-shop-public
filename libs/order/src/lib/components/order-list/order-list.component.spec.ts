import { ComponentFixture, TestBed } from '@angular/core/testing';
import { OrderListComponent } from './order-list.component';
import { UserFacade } from '@demo-shop/user';
import { signal } from '@angular/core';
import { OrderResponse, OrderStatus } from '@demo-shop/api';
import { OrderFacade } from '../../order.facade';
import { provideRouter } from '@angular/router';
import { UTCDate } from '@date-fns/utc';

describe('OrderListComponent', () => {
  let component: OrderListComponent;
  let fixture: ComponentFixture<OrderListComponent>;

  const user = {
    address: {
      street: 'street',
      apartment: 'apartment',
      city: 'city',
      region: 'region',
      zip: 'zip',
      country: 'country',
    },
    email: 'email',
    firstname: 'firstname',
    id: 1,
    lastname: 'lastname',
    phone: 'phone',
  };

  const orders: OrderResponse[] = [
    {
      id: 1,
      userId: 1,
      items: [],
      amount: 0,
      status: OrderStatus.Created,
      created: new UTCDate(2020, 0, 1).toString(),
    },
    {
      id: 2,
      userId: 1,
      items: [],
      amount: 0,
      status: OrderStatus.Created,
      created: new UTCDate(2020, 0, 1).toString(),
    },
  ];

  beforeEach(async () => {
    await TestBed.configureTestingModule({
      imports: [OrderListComponent],
      providers: [
        { provide: UserFacade, useValue: { getCurrentUser: jest.fn().mockReturnValue(signal(user)) } },
        { provide: OrderFacade, useValue: { getSortedByStatusAndDate: jest.fn().mockReturnValue(signal(orders)) } },
        provideRouter([]),
      ],
    }).compileComponents();

    fixture = TestBed.createComponent(OrderListComponent);
    component = fixture.componentInstance;
    fixture.detectChanges();
  });

  it('should create', () => {
    expect(component).toBeTruthy();
  });

  it('should match the snapshot', () => {
    expect(fixture).toMatchSnapshot();
  });
});
