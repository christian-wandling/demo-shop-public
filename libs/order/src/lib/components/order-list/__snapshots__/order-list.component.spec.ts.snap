// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OrderListComponent should match the snapshot 1`] = `
<lib-order-list
  MAX_THUMBNAILS={[Function Number]}
  orders={[Function Function]}
  user={[Function Function]}
>
  <div
    class="mx-auto px-8 max-w-7xl"
  >
    <section
      class="py-8 top-16 sticky z-10 bg-white border-b-2 border-gray-100"
    >
      <h1
        class="font-bold text-gray-700"
      >
        firstname lastname
      </h1>
      <div
        class="text-xs text-gray-500"
      >
        <p>
          street apartment
        </p>
        <p>
          zip city, region
        </p>
        <p>
          country
        </p>
      </div>
    </section>
    <section>
      <ul
        class="divide-y divide-gray-100"
      >
        <li
          class="relative flex flex-col sm:flex-row justify-between gap-x-6 py-5 px-5 cursor-pointer hover:bg-slate-50"
          tabindex="0"
        >
          <lib-order-status
            class="rounded-lg px-4 py-1.5 max-w-28 text-center text-white border border-slate-300 text-xs absolute right-5 top-5 bg-orange-500"
          >
            Created
          </lib-order-status>
          <div
            class="flex min-w-0 gap-x-3"
          >
            <div
              class="min-w-0 flex-auto"
            >
              <p
                class="text-xl mb-4 font-semibold hover:underline"
              >
                Order #1
              </p>
              <div
                class="flex gap-4"
              >
              </div>
            </div>
          </div>
          <div
            class="flex flex-col mt-4 sm:items-end sm:justify-end gap-1"
          >
            <p
              class="text-sm leading-none text-gray-900"
            >
              $0.00
            </p>
            <lib-date-time
              class="text-xs leading-none text-gray-500"
            >
              <time
                datetime="Wed Jan 1 2020 00:00:00 GMT+0000 (Coordinated Universal Time)"
              >
                Jan 01, 2020
              </time>
            </lib-date-time>
          </div>
        </li>
        <li
          class="relative flex flex-col sm:flex-row justify-between gap-x-6 py-5 px-5 cursor-pointer hover:bg-slate-50"
          tabindex="0"
        >
          <lib-order-status
            class="rounded-lg px-4 py-1.5 max-w-28 text-center text-white border border-slate-300 text-xs absolute right-5 top-5 bg-orange-500"
          >
            Created
          </lib-order-status>
          <div
            class="flex min-w-0 gap-x-3"
          >
            <div
              class="min-w-0 flex-auto"
            >
              <p
                class="text-xl mb-4 font-semibold hover:underline"
              >
                Order #2
              </p>
              <div
                class="flex gap-4"
              >
              </div>
            </div>
          </div>
          <div
            class="flex flex-col mt-4 sm:items-end sm:justify-end gap-1"
          >
            <p
              class="text-sm leading-none text-gray-900"
            >
              $0.00
            </p>
            <lib-date-time
              class="text-xs leading-none text-gray-500"
            >
              <time
                datetime="Wed Jan 1 2020 00:00:00 GMT+0000 (Coordinated Universal Time)"
              >
                Jan 01, 2020
              </time>
            </lib-date-time>
          </div>
        </li>
      </ul>
    </section>
  </div>
</lib-order-list>
`;
